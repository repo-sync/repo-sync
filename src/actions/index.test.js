import { getGH, login, getToken, getRepos, createFile } from './'

describe('login redirect', () => {
  const originalLocation = window.location

  beforeEach(() => {
    delete window.location
    window.location = {
      href: ''
    }
  })

  afterEach(() => {
    window.location = originalLocation
  })

  it('should redirect to login', () => {
    window.location = { href: '' }
    login()
    expect(window.location.href).toMatchSnapshot()
  })
})

it('should get token with code', async () => {
  const token = await getToken('testcode')
  expect(token).toMatchSnapshot('testtoken')
})

it('should fail to get token with bad code', async () => {
  const token = await getToken('badcode')
  expect(token).toBeNull()
})

it('should fail to get token without code', async () => {
  const token = await getToken()
  expect(token).toBeNull()
})

it('should get repos with token', async () => {
  const repos = await getRepos('testtoken').data
  expect(repos).toMatchSnapshot()
  const gh = getGH()
  expect(gh.getToken()).toBe('testtoken')
  expect(gh.getUser).toBeCalled()
  expect(gh.listRepos).toBeCalled()
})

it('should fail when getting repos without token', async () => {
  await expect(getRepos()).rejects.toMatchSnapshot()
})

it('should create file', async () => {
  const result = await createFile('owner/repo', '.github/workflows/repo-sync.yml', '<yaml-body>')
  const gh = getGH()
  expect(gh.getUser).toBeCalled()
  expect(gh.getRepo).toBeCalledWith('owner/repo')
  expect(gh.writeFile).toBeCalled()
  expect(gh.writeFile).toMatchSnapshot()
  expect(result).toMatchSnapshot()
})

it('should fail to create file when repo is missing', async () => {
  await expect(createFile(undefined, '.github/workflows/repo-sync.yml', '<yaml-body>')).rejects.toMatchSnapshot()
  const gh = getGH()
  expect(gh.getUser).toBeCalled()
  expect(gh.getRepo).toBeCalledWith(undefined)
})
