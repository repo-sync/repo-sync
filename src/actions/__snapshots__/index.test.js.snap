// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login redirect should redirect to login 1`] = `"https://github.com/login/oauth/authorize?client_id=test_client_id&redirect_uri=&scope=user%20repo"`;

exports[`should create file 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      undefined,
      ".github/workflows/repo-sync.yml",
      "<yaml-body>",
      "Creating .github/workflows/repo-sync.yml",
      Object {
        "committer": Object {
          "email": "committer@example.com",
          "name": "Committer Name",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`should create file 2`] = `
Object {
  "data": Object {
    "content": Object {
      "html_url": "https://github.com/commit-url",
    },
  },
}
`;

exports[`should fail to create file when repo is missing 1`] = `[Error: Failed to create file]`;

exports[`should fail when getting repos without token 1`] = `[Error: Failed to get repos]`;

exports[`should get repos with token 1`] = `undefined`;

exports[`should get token with code: testtoken 1`] = `"testtoken"`;
